### ISSUE TITLE
Feature request: redirect terminal component to stdout

### ISSUE THREAD

**Issue by jsegarraf** [hneemann/Digital#363]
I would like to evaluate the simulation of a given RAM content (with assembler code) in a non-interactive way, without gui. Although this is not trivial, a simple workaround would be to redirect the terminal component to stdout. With such component, I could build a circuit to capture the specific data/signals to evaluate, and write them to a terminal. For instance, the specific values written to a given memory address. Having such sequence redirected to stdout, an external script could validate it.

Possibly, redirecting the keyboard component to stdin could be a complementary feature, also very interesting.

**Comment by SmallRoomLabs** [comment]
I've wrote a module for Digital to do something similar just last week.  It's a TCP socket component that accepts data on a 8-bit parallel port or as 8N1 serial inside the simulation and dumps that data out on a TCP socket that you can connect to with telnet or any software capable of connecting to a TCP port.

The component also takes data that is comes into the socket and present that as a parallel 8-bit value or as a serial stream inside the simulation.

I use this to get more of a TTY/serial terminal feeling for my PDP8 emulation. (And as the serial interface was an integral part of the old PDP8 I want to be able to have that part of the simulation as well, so the internal Terminal/Kbd components doesn't fullfill my needs)

Unfortunately I wrote this as a real Digital component as I wasn't aware of the plug-in architecture. So my sources needs to be added to and compiled into the original java code.

If external comms via a TCP socket would solve your problem I'd happy to give you the code and pre-compiled executables/jars.

**Comment by hneemann** [comment]
@jsegarraf 
That sounds interesting! Can you describe your use case in more detail?
What do you have, what do you want to do, and what information do you need at the end?
Envently one could also extend the existing test case.

**Comment by jsegarraf** [comment]
Students must submit a .hex file with assembler and data for a very simple CPU+RAM circuit. All students must perform the same task (some calculation, string manipulation, etc.), but each one has particular specifications (essentially specific code/data start addresses). So, all .hex are different. I don't think test cases are adequate, since I don't want to test the circuit, but the program in its RAM.

To evaluate each .hex I would need, for instance, to track how a particular memory address is updated, that is, the sequence of data stored there. This could be done by connecting a terminal to the data bus, and enabling it when storing to the desired address. Anyway, the desired data to send to the terminal depends on the particular program to evaluate.

I would like the terminal to print to stdout, so I could redirect it to another script to analyze it.  Ideally, I would like Digital to simulate the circuit automatically (without pressing play), so the gui would not be required. It could be something like:
`java -jar Digital.jar <circuit.dig> | testing_script`

The not-so-elegant part is that the simulation would never stop, so Digital should be killed from the testing_script or via timeout.

Regarding the keyboard-stdin counterpart, it would be similar. In this case, the gui/no-gui could be the selector for the keyboard window/stdin input. Currently I have no use case for it, but I think of having something equivalent to the `cat` command with a circuit connecting a keyboard to a terminal.

Maybe a '--no-gui' parameter in Digital could be useful for this and other features.

**Comment by hneemann** [comment]
It seems to me that you need above all the possibility to write a kind of unit tests for an assembly program. Hmm! I have to think about it.

**Comment by hneemann** [comment]
@jsegarraf 
I suspect that your test cases have to be very flexible, and therefore at least a script language is required.
What if the tests were written in Java?

**Comment by hneemann** [comment]
@jsegarraf 
This is how a java JUnit test could look like:

``` java
public class UnitTestTest extends TestCase {

    public void testProcessor() throws Exception {
        // the file containing the processor
        File processor = new File(Resources.getRoot(),
                "../../main/dig/processor/Processor.dig");

        // the hex file containing the program
        File hexFile = new File(Resources.getRoot(),
                "programs/fibonacci.hex");

        // import the hex file
        DataField program = Importer.read(hexFile, 16);

        // run the program
        RAMInterface ram = new UnitTester(processor)    // load processor
                .writeDataTo(pm -> pm instanceof ROM && pm.getDataBits() == 16, program)
                                                        // write program to the 16-Bit rom
                .runToBreak()                           // run program to break point
                .getRAM(v -> v instanceof RAMDualPort); // get the data ram of processor

        // get the content of the data ram
        DataField ramContent = ram.getMemory();

        // check if address 0 contains the value 610 ( the fifteenth Fibonacci number )
        assertEquals(610, ramContent.getDataWord(0));
    }
}
```

**Comment by jsegarraf** [comment]
I agree that some language for such test cases is needed. Although java would be fine for the general case, it is too much for me, since I just need to verify a sequence such as "369c". This would be very easy to perform from a script, and nevertheless I would need some external scripting to write the results to import to other platforms (Moodle).

So, I would still prefer to modify my circuit to add some external output, and then manage that output outside Digital. Probably, any user would be able to perform a circuit modification to get some output, but not all of them would be willing to program it using java.

**Comment by hneemann** [comment]
Here's another idea: Why not write a single small Java program that is suitable for your needs and does exactly what you need for your evaluations.
[Here](https://github.com/hneemann/DigitalTesting) you can find a small example which is started with three arguments: The circuit to open, the hex file to load into memory, and a filename under which to store the final RAM content.
Once compiled, it can be started from a script and the results can also be evaluated from that script.

**Comment by jsegarraf** [comment]
The idea seems simple enough to work for my use case.
I plan to use these tests around mid 2020, so it may take some time until I can actually test it.

**Comment by hneemann** [comment]
closed in e96d91742e44521f1b128e8221fdb8f6c8e7a980

@jsegarraf 
Please reopen the issue if any problems occur.
